Adding binders for STL containers

lyskov opened the pull request:
First draft for STL container binders. @wjakob  could you please let me know if general structure is ok or if you have any suggestions. And i will add more binders and functionality once we establish general structure.

Thanks,

wjakob commented on include/pybind11/stl_binders.h:
```diff
@@ -0,0 +1,184 @@
+template<typename T>
+constexpr bool has_insertion_operator()
+{
+	using namespace has_insertion_operator_implementation;
```
MSVC2015's SFINAE implementation probably can't handle such a fancy type trait class.

wjakob commented on include/pybind11/stl_binders.h:
```diff
@@ -0,0 +1,184 @@
+
+		// Python friendly bindings
+		#ifdef PYTHON_ABI_VERSION // Python 3+
+			cl.def("__bool__",    [](Vector &v) -> bool { return v.size(); }); // checks whether the container has any elements in it
```
Add ``!=0`` instead of bool cast, which causes warnings on some compilers

wjakob commented on include/pybind11/stl_binders.h:
```diff
@@ -0,0 +1,184 @@
+		});
+
+		cl.def("__setitem__", [](Vector &v, SizeType i, T const & t) {
+            if (i >= v.size()) throw pybind11::index_error();
```
minor: indentation issue? (tabs vs spaces?)

wjakob commented on include/pybind11/stl_binders.h:
```diff
@@ -0,0 +1,184 @@
+
+		// Modifiers
+		cl.def("clear",                                     &Vector::clear,     "clears the contents");
+		cl.def("push_back",    (void (Vector::*)(const T&)) &Vector::push_back, "adds an element to the end");
```
A general thought: since you're implementing the Python interfaces (__len__, __setitem__), it probably makes sense to implement functions that behave like a Python ``list`` in a duck-typing context. So ``append`` instead of ``push_back``, etc.

lyskov replied:
Yeah, i thought about this as well. The reason for push_back (and other C++ style functions) is make container classes to behave as C++ counterparts (i.e. principle of least surprise). This will allow users to use resources like http://en.cppreference.com to see what methods are available and their properties.

But i see benefits of having containers to be more 'Pythonish' as well. How about if we have both C++ and Python-style methods implemented? (like __len__ and size()), will you be ok with that?

wjakob replied:
I'd be greatly in favor of only implementing the Pythonic methods (this also what Boost.Python does). I suggest returning the ``class_`` instance in the ``vector_indexing_suite`` code so that the caller could still  add the C++ methods if they are strictly required in a specific project.

lyskov replied:
I thought this over and i agree that it does not make sense to bind both Python and C++ methods which will just create clutter and increase binding size. So i am going to comment all C++ related methods that does have good counterparts in Python.

Re returning `class_` instance: - good idea, adding this now...

lyskov replied:
re Re returning `class_` instance: since binder itself is just a call to class contractor then we can't return any value from it. Instead i added `class_` instance as member and user will be able to access it if needed.

wjakob commented on include/pybind11/stl_binders.h:
```diff
@@ -0,0 +1,184 @@
+
+NAMESPACE_BEGIN(pybind11)
+
+template<typename T>
```
maybe simpler to have a ``has_equality_operators`` template that tests for both ``==` and ``!=``

lyskov replied:
Hmmm sure, i can simplify this... but what about the cases when user type T have '==' but does not have '!=' or otherwise?

wjakob replied:
I think that would be a pretty broken implementation of the underlying type, not something that we need to worry about in pybind11.

wjakob commented:
Hi Sergey,

looks like a great start! I added some inline comments to your PR.

-Wenzel

lyskov commented:
@wjakob i addressed some of the comments and replied to others (see last two comments for outdated diff's). 

Re MSVC build: i added ifdef's to disable insertion-operator-detection implementation for now. (but please let me know if you know of way to do so in way MSVC accept). Windows build still has not compiled and reported error seems to be the else-clause for std::is_copy_constructible, any idea how to fix this?

I also added some 'Python-style' modifiers and they require Python ValueError exception so i added `PYBIND11_RUNTIME_EXCEPTION(value_error)`, but i think i missed something because this exception does not seems to be synced to Python::ValueError. Could you please advise me of how to fix this?

Thanks,

lyskov commented:
never mind the `value_error` question - i pushed commit that fix this

wjakob commented on include/pybind11/pybind11.h:
```diff
@@ -422,6 +422,7 @@ class cpp_function : public function {
             }
         } catch (const error_already_set &)      {                                                 return nullptr;
         } catch (const index_error &e)           { PyErr_SetString(PyExc_IndexError,    e.what()); return nullptr;
+        } catch (const value_error &e)           { PyErr_SetString(PyExc_ValueError,    e.what()); return nullptr;
```
Could you also add a note about this new kind of exception to ``docs/advanced.rst``?

wjakob commented on example/example17.cpp:
```diff
@@ -0,0 +1,27 @@
+#include <pybind11/stl_binders.h>
+
+class A
+{
```
pybind11 code has the braces on the same line (sorry, I'm pedantic ;))

lyskov replied:
No problem and totally understood! I will try to fix all instances of these...

wjakob commented on example/example17.cpp:
```diff
@@ -0,0 +1,27 @@
+
+void init_ex17(py::module &m)
+{
+	pybind11::class_<A>(m, "A");
```
Indentation issue?

wjakob commented on example/example17.py:
```diff
@@ -0,0 +1,24 @@
+from example import VectorInt, VectorA
+
+v_int = VectorInt(2)
+print( v_int.size() )
```
pybind11 code doesn't have extra spaces before/after parentheses (sorry, pedantic ;))

wjakob commented on include/pybind11/stl_binders.h:
```diff
@@ -0,0 +1,226 @@
+			cl.def("__nonzero__", [](Vector &v) -> bool { return v.size() != 0; }); // checks whether the container has any elements in it
+		#endif
+
+		cl.def("__getitem__", [](Vector const &v, SizeType i) {
```
Slice based __getitem__ and __setitem__ would be important to have. See ``example/example6.cpp`` for reference.

wjakob commented:
Looks like ``__iter__`` is still missing? (that's the most important function IMO). See example6 for a simple two-liner.

wjakob commented:
Regarding the Visual Studio compiler issue: the way I solve these is with a MSVC install in a virtual machine, where I build up a little snippet (outside of pybind11) until it breaks, followed by a search for workarounds. This usually yields something workable after 2-3 iterations.

lyskov commented:
@wjakob i pushed changes that i think addresses most of the issues/suggestions. Please see change log for details and let me know if you have further suggestions. So i unless there is other things the only remaining issue is Windows build and merging of == and != operators detection. I will address this when i figure out how to test MSVC build.

Thank you for the tip about MSVC - i was not aware it is now available as free download. I will try to setup VM and see if i can make this work.

wjakob commented on include/pybind11/stl_binders.h:
```diff
@@ -0,0 +1,260 @@
+				return seq;
+			});
+	}
+	template<typename U = T, typename std::enable_if< !std::is_default_constructible<U>{} >::type * = nullptr>
```
I have not seen this syntax before (``std::enable_if< !std::is_default_constructible<U>{} >::type``) and use the following elsewhere in the codebase: ``std::enable_if< !std::is_default_constructible<U>::value>::type``. Just mentioning it in case it makes any difference for MSVC

lyskov replied:
Right, that was indeed the problem with this part of the code. Idea was that {} will trigger struct to be constructed and then we relay on bool() operator to convert struct to result value. I guess MSVC version for some reason fail to initialize struct. Anyway: i converted all this code to use ::value instead.

wjakob commented:
One more suggestion (the last I think): there is nothing that particularly needs ``vector_binder`` to be a class. The approach that I've generally followed is to make this kind of code a function in the ``pybind11`` namespace, while isolating the messy enable_if<> bits in ``pybind11::details``. That way, you can also give it the return value ``class_<std::vector<type, allocator, ..>>`` etc.

lyskov commented:
Right, i made `vector_binder` a class exactly to hide messy enable_if's. Good idea about putting them all in `details` instead, i will look it up!

wjakob commented:
Great, thank you -- I look forward!

lyskov commented:
@wjakob i refactored the code so `vector_binder` is now a function instead of class.

Also: i fixed the Windows build so it now compile without errors, however test itself seems to be failing but without any message, any idea what might be wrong?

wjakob commented:
Some of the testcases seem to be giving different results than the reference data in the ".ref" files (?)

wjakob commented:
I did a pass over your code and merged it with some adaptations -- can you take a look?

lyskov commented:
@wjakob looks great! A few notes:

- `__bool__` will only work in Python3+ for Python2 we need to use `__nonzero__` - will you be open if i back ported the #ifdef solution?

- re bindings for `max_size reserve capacity shrink_to_fit clear swap ` while they clearly not a 'Python' functions they could be very useful in some circumstances. Do you have a strong opinion of not binding them? If yes, how about binging for only `reserve shrink_to_fit`? Rational is that such bindings will allow user to effectively manage memory if such need arise etc.

- what about the windows build? What the issue that prevent test to run?

Thanks,

wjakob commented:
Regarding your first point, see https://github.com/pybind/pybind11/commit/d1bfc4e0f8acc9f65bc7d19de5ff6ecf854ebf93.

I'd rather leave out the C++ memory management stuff, which seems very un-pythonic to me. The resulting type should behave exactly like a Python list.

I changed the SFINAE code quite a bit until it compiled on all platforms. Currently, it's passing all tests on my end (and the Travis/AppVeyor CI bots).

lyskov commented:
oh, i missed d1bfc4e0f8acc9f65bc7d19de5ff6ecf854ebf93 - great!

Hmm, last iteration of original PR compile fine, i think it might run into some runtime problem. Oh well, i guess that ship is sailed now.

re C++ function: - yes, i see your point. And since we return related `class_` instances users can expand this if needed.

Future: since now framework is established, will you be open for future PR adding binding for other STL containers?

wjakob commented:
Yes, other STL containers would be fantastic!
